### ISSUE TITLE
ZStream.aggregateAsyncWithin retains objects from aggregation function

### ISSUE THREAD

**Issue by reimai** [zio/zio#6613]
In my project I'm using ZStream to essentially fold big messages into a small one using aggregateAsyncWithin, and write these to storage. I've noticed that every time a schedule loses a race, all aggregated chunks are held in memory until that schedule step finishes. 

In some scenarios this leads not only to a huge memory footprint, but to oom (if there is a lot of unread messages, fold always completes faster than schedule).     

Here is a demo project, it goes to oom in 5 pulls: 
https://github.com/reimai/zio_aggregate_oom/blob/master/src/main/scala/Main.scala

**Comment by adamgfraser** [comment]
This does not appear to be a problem on ZIO 2.0. Ran the following code on `series/2.x` and it appeared to be going strong as long as I had patience to view console output.

```scala
package zio.stream

import zio._

object Main extends ZIOAppDefault {

  private val ones = Array.fill[Byte](256 * 1024)(1)
  case class InboxMessage(bytes: Array[Byte] = ones.clone())

  val run =
    ZStream
      // here I would expect at most 2 InboxMessages in the heap, one in handoff and one in transducer's fold
      .fromIterator(Iterator.continually(InboxMessage()), 1)
      .aggregateAsyncWithin(
        // a batch of 1024 InboxMessages would be ~250mb
        ZSink.fold[InboxMessage, Int](0)(_ < 1024) { case (acc, _) =>
          acc + 1
        },
        // but I'm seeing multiple of 250mb batches in a heap, retained by race-losing schedules' fibers observers
        Schedule.spaced(1.hour)
      )
      .mapZIO { batch =>
        // oom after 5 batches with a 2gb heap
        Console.printLine(s"got $batch")
      }
      .runDrain
}
```

**Comment by reimai** [comment]
That's great. Although 2.0 is still a release candidate. Could we expect fixes in 1.0? I see the ZStream has been completely rewritten, so, sadly, no simple port is possible.

**Comment by MaksGS09** [comment]
> This does not appear to be a problem on ZIO 2.0. Ran the following code on `series/2.x` and it appeared to be going strong as long as I had patience to view console output.
> 
> ```scala
> package zio.stream
> 
> import zio._
> 
> object Main extends ZIOAppDefault {
> 
>   private val ones = Array.fill[Byte](256 * 1024)(1)
>   case class InboxMessage(bytes: Array[Byte] = ones.clone())
> 
>   val run =
>     ZStream
>       // here I would expect at most 2 InboxMessages in the heap, one in handoff and one in transducer's fold
>       .fromIterator(Iterator.continually(InboxMessage()), 1)
>       .aggregateAsyncWithin(
>         // a batch of 1024 InboxMessages would be ~250mb
>         ZSink.fold[InboxMessage, Int](0)(_ < 1024) { case (acc, _) =>
>           acc + 1
>         },
>         // but I'm seeing multiple of 250mb batches in a heap, retained by race-losing schedules' fibers observers
>         Schedule.spaced(1.hour)
>       )
>       .mapZIO { batch =>
>         // oom after 5 batches with a 2gb heap
>         Console.printLine(s"got $batch")
>       }
>       .runDrain
> }
> ```

Hi!
@adamgfraser I'm faced with similar OOM with `aggregateAsyncWithin ` on ZIO 2.0.0. Also I'm catching OOM with above example.
The OOM occurs on -Xmx512m after ~ 6 min of run.
Could you look at this again please?

**Comment by reimai** [comment]
Can confirm, I've [updated my example to zio 2.0.0](https://github.com/reimai/zio_aggregate_oom/blob/master/src/main/scala/Main.scala) and it still fails with an oom. Also tried it locally with the [recent leak fix](https://github.com/zio/zio/pull/7069), result is the same.

**Comment by adamgfraser** [comment]
Yes will do.

**Comment by reimai** [comment]
Actually, oom in zio 2 is different. It does not need a huge timeout in aggregation function, and heap content is different, full of ChannelExecuttors. I've [updated](https://github.com/reimai/zio_aggregate_oom/blob/master/src/main/scala/Main.scala) my example a bit (chunk size = 1, smaller timeout and batches). Ooms on the [version](https://github.com/zio/zio/pull/7234) with the recent leak fix.

**Comment by reimai** [comment]
Yeah, https://github.com/zio/zio/issues/7235 fixes it for zio2, thanks.
Still fails on series/1.x

**Comment by hearnadam** [comment]
We are not planning any more work on ZIO 1.
